blockchain/chain/src/store/index.rs

```diff
@@ -17,6 +17,70 @@ const DEFAULT_CHAIN_INDEX_CACHE_SIZE: NonZeroUsize =
 /// Configuration which sets the length of tipsets to skip in between each cached entry.
 const SKIP_LENGTH: ChainEpoch = 20;
 
+// this module helps speed up boot times for forest by checkpointing previously seen tipsets from snapshots.
+pub(super) mod checkpoint_tipsets {
+    use cid::Cid;
+    use forest_blocks::TipsetKeys;
+    use once_cell::sync::Lazy;
+    use std::collections::HashMap;
+    use std::str::FromStr;
+
+    macro_rules! add_calibnet {
+        ($map: ident, $key_hash:expr) => {
+            $map.insert(
+                $key_hash,
+                // calibnet genesis tipset keys
+                TipsetKeys::new(vec![Cid::from_str(
+                    "bafy2bzacecyaggy24wol5ruvs6qm73gjibs2l2iyhcqmvi7r7a4ph7zx3yqd4",
+                )
+                .unwrap()]),
+            );
+        };
+    }
+
+    macro_rules! add_mainnet {
+        ($map: ident, $key_hash:expr) => {
+            $map.insert(
+                $key_hash,
+                // mainnet genesis tipset keys
+                TipsetKeys::new(vec![Cid::from_str(
+                    "bafy2bzacecnamqgqmifpluoeldx7zzglxcljo6oja4vrmtj7432rphldpdmm2",
+                )
+                .unwrap()]),
+            );
+        };
+    }
+
+    // The hashes for these checkpoints is obtained by passing the tipset keys' cids (tipset.cids())
+    // through a blake2b hasher.
+    type TipsetKeyHash = &'static str;
+    type GenesisTipsetCids = TipsetKeys;
+    // A map of validated checkpoints mapping to their genesis tipkeys
+    type TipsetCheckpointsRegistry = HashMap<TipsetKeyHash, GenesisTipsetCids>;
+    // Represents a static map of validated tipset hashes which helps to remove the need to validate the tipset
+    // back to genesis if it has been validated before, thereby reducing boot times.
+    pub(super) static TIPSET_CHECKPOINTS: Lazy<TipsetCheckpointsRegistry> = Lazy::new(|| {
+        let mut map = HashMap::new();
+        // The numeric suffix in the variable name is the height of the snapshot.
+        // NB: Add desired tipset checkpoints below this by using RPC command: forest-cli chain tipset-hash <cid keys>
+        const CALIBNET_CHECKPOINT_1405400: TipsetKeyHash = "7930ad8bf32b35314b3bc47b9e25249af8ec6ba7f5544c05e8b5bd3b3ec09f76df8bd2278f9b318badf1a08d0a468abd55130465c6c55f99e67badc0e614ca79";
```

Comment:
1405400 does not look like the correct height in rebooted calibnet

Reply:
yep, am currently rebasing and integrating changes from recent calibnet snapshot

Reply:
Added current snapshot's tipset hash at height: 41000 below